Project for domain specific datasets

burtenshaw opened the pull request:
This PR implements an application to help people collaborate with domain experts when building datasets. It uses a UI deployed in a Huggingface space to define the prompts and tasks for generating a synthetic dataset with [distilabel](https://github.com/argilla-io/distilabel).

## EXTENSIBLE WORKFLOW

We have two spaces:
- The project welcome https://huggingface.co/spaces/argilla/domain-specific-datasets-welcome that helps users to create a dataset repo and resources
- The project specific space that helps users to manage their synthetic data projects https://huggingface.co/spaces/argilla/domain-specific-datasets-template

The workflow should be self contained within the applications.

davanstrien commented:
Think this is a really nice idea! A couple of higher-level questions

IUC the seed space doesn't need to be duplicated by the user. They can use the Space you created and create their initial seed dataset? I think this is a great flow; even getting people to develop these seed datasets is nice, IMO. 

A couple of questions:
- Is there a strong preference for using the mistral API for the generations? If not, it could reduce the cost of creating the dataset if we can use the serverless Inference APIs (https://huggingface.co/blog/inference-pro). This would make the dataset free (or ~$9 month if they want higher rate limits)
- Related to the above, it could potentially make sense to create a second space for running the scripts. This would take a bit more work (I'm happy to do/help with this), but this could lower the barrier quite a bit for users and make it possible for domain experts to directly create the datasets. 

It would require a bit of extra work to do the second Space, so I think it might make more sense to share the current flow and then iterate toward the no-code version if there seems to be enough interest (I'm happy to help with that work). I guess you want to wait to share this more widely until the 1.0 `distilabel` launch?

davidberenstein1957 commented on domain-specific-datasets/README.md:
```diff
@@ -0,0 +1,17 @@
+# Domain Specific Dataset Project
+
+This is project to bootstrap the creation of domain-specific datasets for training models. The goal is to create a set of tools that help users to collaborate with domain experts.
```
I think a step by step example would be more useful in this case. Even something like a video of you, or anyone else, walking through the idea and steps one-by-one might be even better.

davidberenstein1957 commented on domain-specific-datasets/README.md:
```diff
@@ -0,0 +1,17 @@
+The seeding space: https://huggingface.co/spaces/argilla/domain-specific-seed for the domain expert.
+The demo dataset: https://huggingface.co/datasets/argilla/farming we've created
+The argilla space: https://huggingface.co/spaces/argilla/farming
+The pipeline code: https://github.com/argilla-io/distilabel-workbench/tree/main/projects/farming
```
This is a good to have as an example.

davidberenstein1957 replied:
I think the distilabel pipeline (using random LLMs, when to use GPU, when not, where to plugin API keys etc.) is going to be difficult for most users, when not explained well.

davidberenstein1957 commented on domain-specific-datasets/app/DATASET_README_BASE.md:
```diff
@@ -0,0 +1,6 @@
+# Domain Dataset Grower
```
I think the wordplay is fun but maybe due to the seed, farming etc is a bit difficult to get what is the core of the project and what is the fun example with a pun, when you try to learn what the project is about.

davidberenstein1957 commented on domain-specific-datasets/app/app.py:
```diff
@@ -0,0 +1,124 @@
+)
+from pipeline import serialize_pipeline
+
+st.markdown("# 🧑‍🌾 Domain Data Grower")
```
Can't we combine these markdowns?

davidberenstein1957 commented on domain-specific-datasets/app/app.py:
````diff
@@ -0,0 +1,124 @@
+    instructions = f"""
+        Execute the following command to generate a synthetic dataset from the seed data:
+        ```bash
+        git clone https://huggingface.co/{hub_username}/{project_name}
````
should we add the install distilabel part here too?

davidberenstein1957 replied:
Perhaps just storing the pipeline code somewhere and explaining people then can copy-paste and change it might be very useful to simplify the process.

davidberenstein1957 commented on domain-specific-datasets/app/defaults.py:
```diff
@@ -0,0 +1,17 @@
+
+with open(SEED_DATA_PATH) as f:
+    DEFAULT_DATA = json.load(f)
+DEFAULT_DOMAIN = DEFAULT_DATA["domain"]
```
do we want to use default data if people are supposed to clone and change this?

burtenshaw replied:
The idea would not be to clone the seeding space. Instead, they would create a dataset from a single space (probably hosted on the DIBT org).

Personally, I find that the default data help to explain: What a domain, perspective, topic, and system prompt are.

davidberenstein1957 commented on domain-specific-datasets/app/domain.py:
```diff
@@ -0,0 +1,85 @@
+load_dotenv()
+
+# Application description used for SelfInstruct
+APPLICATION_DESCRIPTION = f"""You are an AI assistant than generates queries around the domain of {DEFAULT_DOMAIN}.
```
Perhaps we want people to be able to change this too?

davidberenstein1957 commented on domain-specific-datasets/app/app.py:
```diff
@@ -0,0 +1,124 @@
+st.header("📚 Examples")
+
+questions_answers = []
+
```
should we consider adding calling the API with an example the data in order to have an idea and allow for prompt engineering within the streamlit while they are fine-tuning their input?

davidberenstein1957 commented on domain-specific-datasets/app/pipeline.py:
```diff
@@ -0,0 +1,109 @@
+            data=[{"input": term} for term in terms],
+            batch_size=64,
+        )
+        base_llm = MistralLLM(
```
maybe this is too specific and we should emphasize people can use other LLMs than mistral?

davidberenstein1957 commented on domain-specific-datasets/app/pipeline.py:
```diff
@@ -0,0 +1,109 @@
+    pipeline_config_path: str = "pipeline.yaml",
+):
+    terms = create_topics(topics, perspectives)
+    with Pipeline("farming") as pipeline:
```
naming might be too general too? ("farming")

davidberenstein1957 commented on domain-specific-datasets/app/pipeline.py:
```diff
@@ -0,0 +1,109 @@
+        evol_instruction_complexity = EvolInstruct(
+            name="evol_instruction_complexity",
+            llm=base_llm,
+            num_evolutions=2,
```
should we make this configurable?

davidberenstein1957 commented on domain-specific-datasets/app/pipeline.py:
```diff
@@ -0,0 +1,109 @@
+            name="keep_columns",
+            columns=["model_name", "evolved_questions", "domain_expert_answer"],
+        )
+
```
do we perhaps want to add `EvolQuality` here? 
Also, potentially something like PairRM would be nice? I
Within my pipeline I was using `EvolQuality` with 3 evols and `PairRM` after so we would directly have a preference dataset too.

davidberenstein1957 commented on domain-specific-datasets/app/pipeline.py:
```diff
@@ -0,0 +1,109 @@
+        to_argilla = TextGenerationToArgilla(
+            name="text_generation_to_argilla",
+            dataset_name=argilla_dataset_name,
+            dataset_workspace="admin",
```
maybe this should be configurable too?

davidberenstein1957 commented on domain-specific-datasets/app/infer.py:
```diff
@@ -0,0 +1,16 @@
+import os
```
Things I mentioned earlier about inference and potentially difficult API/LLM usage etc.

A lot of community members don't seem to have money for LLM API access.

davidberenstein1957 commented on domain-specific-datasets/app/pipeline.yaml:
```diff
@@ -0,0 +1,471 @@
+distilabel:
```
I think pushing the yaml approach is cool but more useful for more advanced users.

burtenshaw replied:
We use the yaml to share and execute the pipeline.

davidberenstein1957 commented on domain-specific-datasets/scripts/embed.py:
```diff
@@ -0,0 +1,46 @@
+from datasets import Dataset, load_dataset
```
Perhaps this is easier and limits the exposed logic?

https://docs.argilla.io/en/latest/practical_guides/create_update_dataset/vectors.html#add-sentence-transformers-vectors

burtenshaw replied:
Yes. We should do this:

```python
from argilla.client.feedback.integrations.sentencetransformers import SentenceTransformersExtractor

dataset = ... # FeedbackDataset or RemoteFeedbackDataset

tde = SentenceTransformersExtractor(
    model="TaylorAI/bge-micro-v2",
    show_progress=True,
)

dataset = tde.update_dataset(
    dataset=dataset
    fields=None # None means using all fields
    update_records=True # Also, update the records in the dataset
    overwrite=False # Whether to overwrite existing vectors
)
```

davidberenstein1957 reviewed (commented):
Looking good already🥳

 I've left some comments which mostly touch upon making it easier for the large public to understand what is going on and be able to participate.

burtenshaw commented:
> IUC the seed space doesn't need to be duplicated by the user. They can use the Space you created and create their initial seed dataset? I think this is a great flow; even getting people to develop these seed datasets is nice, IMO.
Correct.

> * Is there a strong preference for using the mistral API for the generations? If not, it could reduce the cost of creating the dataset if we can use the serverless Inference APIs (https://huggingface.co/blog/inference-pro). This would make the dataset free (or ~$9 month if they want higher rate limits)
There's no attachment to Mistral's API. However, some domains, including farming, require models of a sufficient size to get high quality diverse data. But maybe that's a part of the exercise that the user can iterate over.

> * Related to the above, it could potentially make sense to create a second space for running the scripts. This would take a bit more work (I'm happy to do/help with this), but this could lower the barrier quite a bit for users and make it possible for domain experts to directly create the datasets.
Interesting, I played around with this initially so the :seedling: button just ran the pipeline. This is definitely something we could try out, but I think it would be useful to give the command in a code fence as dropdown option. 
 
> It would require a bit of extra work to do the second Space, so I think it might make more sense to share the current flow and then iterate toward the no-code version if there seems to be enough interest (I'm happy to help with that work). I guess you want to wait to share this more widely until the 1.0 `distilabel` launch?
We don't need to share before or with the 1.0 launch, but it would be nice to get it out soon. 

I'll make these changes:

- [ ] Use inference endpoints with option to select model, default to free model
- [ ] Try no code option for the running the pipeline in the space

burtenshaw commented:
@davanstrien I've refactored the application and responded to your comments, but there's still some work to do:

## Changes

- Improved prose to explain the steps
- 'Add topic' etc buttons so the app is more concise/flexible
- Functionality to run the pipeline within the space
- Support for inference endpoints

![Screenshot from 2024-04-17 07-15-08](https://github.com/huggingface/data-is-better-together/assets/19620375/d2c4fb2c-d36e-48c2-b74c-9aec29d089c7)
![Screenshot from 2024-04-17 07-20-45](https://github.com/huggingface/data-is-better-together/assets/19620375/27fa8554-5dae-4975-9564-cbc0073a2109)
![Screenshot from 2024-04-17 07-36-59](https://github.com/huggingface/data-is-better-together/assets/19620375/daa9e145-619c-49c9-9f66-c4899981d951)

## Remaining Improvements

- [ ] Implement support and interface for freed inference APIs. Ideally there would be a list of available models.
- [ ] The application is becoming very long and I think it will be unpractical to use. We should refine the structure with pages, columns, etc to reflect the workflow that the participants should follow.
- [ ] Currently the application pushes records to a single Argilla instance. Ideally, the space would allow users to define their own Argilla instance's credentials.
- [ ] Once records are labelled in Argilla, they need to be pulled and pushed to the dataset

ignacioct commented:
Hi all! Joining the work on this one.

ignacioct commented:
Regarding the layout of the page, as briefly discussed with @burtenshaw, I've implemented a simple tab structure for the project details, the domain expert, the domain perspectives, the domain topics, and the examples. I think they work well for getting the user to move sequentially in between them the first times, and once the know how to use the app, to go straight to whatever they'd like to configure. We could enforce this behavior with a brief sentence in the introduction. Tell me what you think of this change, and if not we revert and follow some other path.

<img width="748" alt="Screenshot 2024-04-17 at 12 02 12" src="https://github.com/huggingface/data-is-better-together/assets/56955040/d22c508c-5164-47e6-842b-65b8fc2678d3">

burtenshaw commented:
> Regarding the layout of the page, as briefly discussed with @burtenshaw, I've implemented a simple tab structure for the project details, the domain expert, the domain perspectives, the domain topics, and the examples. I think they work well for getting the user to move sequentially in between them the first times, and once the know how to use the app, to go straight to whatever they'd like to configure. We could enforce this behavior with a brief sentence in the introduction. Tell me what you think of this change, and if not we revert and follow some other path.

This looks rad @ignacioct.

@davidberenstein1957 @davanstrien Mentioned making the workflow clear to users. Personally, I'm a bit worried about the process taking so long that the user would want to work async, or restart from certain points.

Here are some suggestions on those lines:

- What if we add a landing page for project details and project explanation? This could end with creating the dataset
- Then a second page for everything else that relates to seed creation (domain expert, topics, perspectives, examples), this could end with a button that adds the seed data and pipeline yaml to the dataset repo
- Then a third page for everything relating to running the pipeline with two tabs (local, on the space). This page could then have a dataset url box to start from the dataset so that the user can work async, and just restart from seed data that they saved in the dataset repo.
- We could use the side bar to explain the phases.

This is a chunk of work so I'm happy to take some on.

davanstrien commented:
> @davidberenstein1957 @davanstrien Mentioned making the workflow clear to users. Personally, I'm a bit worried about the process taking so long that the user would want to work async, or restart from certain points.
> 
> Here are some suggestions on those lines:
> 
>     * What if we add a landing page for project details and project explanation? This could end with creating the dataset
> 
>     * Then a second page for everything else that relates to seed creation (domain expert, topics, perspectives, examples), this could end with a button that adds the seed data and pipeline yaml to the dataset repo
> 
>     * Then a third page for everything relating to running the pipeline with two tabs (local, on the space). This page could then have a dataset url box to start from the dataset so that the user can work async, and just restart from seed data that they saved in the dataset repo.
> 
>     * We could use the side bar to explain the phases.
>  

That sounds good to me. I agree it doesn't make sense to have one thing that takes a very long time to run. IUC creating the seed dataset is fairly quick but the process of running the generation pipeline will take a bit more time. One option for this if it becomes difficult to have multiple users running that from the same app is to either encourage people to clone the app to their own Space i.e. similar to https://huggingface.co/spaces/autotrain-projects/autotrain-advanced or to create a new space via the API to run the code (the first is probably easier/safer).

burtenshaw commented:
@davanstrien This sounds good. How would you encourage cloning effectively?

That means our remaining steps are:

## Readme:
- [ ]  Visual Overview of workflow
- [ ]  Explain distilabel concepts
- [ ]  Explain llm definition
- [ ]  Explain inference endpoints and inference api

## App:
- [ ] Add a page to integrate with Argilla.

@davidberenstein1957 @dvsrepo @davanstrien What do you think about releasing it early next week?

burtenshaw commented:
I made this visual overview https://www.figma.com/file/ldbVfPBrXFRP4hNbCu3trL/Dataset-Grower-Workflow?type=whiteboard&node-id=603%3A109&t=B6lvnCnQDdxCI1wl-1

I think we could put it in the README, and maybe make a video that talks through it.

burtenshaw commented:
We will go for the two app approach suggested by @davanstrien :

- DIBT/domain-specific-datasets - [one instance] @ignacioct 
    - create a dataset
    - create an argilla space
    - create the project app with the above constants
- {user}|{org}/{domain} - [instance per project] @burtenshaw 
    - define domain specific seed 
    - push seed data to dataset repo
    - show code to copy and run the pipeline locally
    - button to run the pipeline on space

davanstrien commented:
This sounds great! Agree a video would be very nice. I also plan to do the same with the other DPO project.

burtenshaw commented:
@davidberenstein1957 I was thinking more about making this an approachable initiative, and thought about defining a range of levels for a project. Because building a alignment scale datasets might be outside some community members' available time. How about these as options:

- :mouse: Generate and label an evaluation dataset to benchmark existing models on your domain.
- :rabbit: Generate an SFT dataset 
- :dog: Generate a DPO/ ORPO dataset
- :lion: Align a language model using ORPO

ignacioct commented:
Hi all! Just pushed some commits in which the functionality of the parent app gets almost fully schemed. However, I have a question regarding weird behavior when creating the dataset. I'm using this function to create an empty dataset:

```
def setup_dataset_on_hub(repo_id, hub_token):
    # create an empty dataset repo on the hub
    hf_api.create_repo(
        repo_id=repo_id,
        token=hub_token,
        repo_type="dataset",
    )
```

And I've been getting this error all morning: 

![Screenshot 2024-04-22 at 12 48 16](https://github.com/huggingface/data-is-better-together/assets/56955040/b5a708b3-e51e-409e-b81a-72253fde2b90)

Is it something I'm doing wrong or did I just got an timeout?

davanstrien commented:
> Is it something I'm doing wrong or did I just got an timeout?
We had some issues with the Hub earlier, so I suspect that caused it. Hopefully should be fixed now.

burtenshaw commented:
@davanstrien It would be great if you could take a look at the how we're interacting with the hub and inference endpoints. And maybe suggest some improvements and/or documentation.

ignacioct commented on domain-specific-datasets/README.md:
```diff
@@ -0,0 +1,81 @@
+
+> **Domain perspectives** are the perspectives that the domain expert wants to include in the dataset. For example, if the domain is farming, the domain perspectives could be "farmer", "agricultural scientist", "agricultural economist", etc.
+
+### 4. Generate the dataset
+
+Next we can move on to generating the dataset from the seed data.
+
+![Run the pipeline](https://raw.githubusercontent.com/huggingface/data-is-better-together/3ac24642454764c8c7d56f0ffdd1a134c1cd37b1/domain-specific-datasets/assets/pipeline.png)
+
```
I'll add here the same info we have on models that people can use in the child app.

ignacioct reviewed (approved):
LGTM! I'll add that detail to the README file

davanstrien commented:
@burtenshaw let me know when you are ready to merge this initial version and we can follow up with additional PRs :)

burtenshaw commented:
@davanstrien I think we're ready to merge. Some remaining comments are:

- Try the space here: https://huggingface.co/spaces/argilla/domain-specific-datasets-welcome
- If you want to review fixes in the project app, you'll need to restart the project from the welcome app to get the changes. :face_with_spiral_eyes:
- Should we migrate the spaces to the DIBT org? 
- @burtenshaw Next step is to make a video of the project workflow
- @dvsrepo @davidberenstein1957 share on monday.

:smile:

burtenshaw commented:
@davanstrien @ignacioct 

I took another look at this end-to-end by testing the pipeline on some mock domains. I wasn't satisfied with how the codeless runtime interacted with distilabel. In short, for these reasons:

- If the pipeline hit an error (caused by configuration, infrastructure, or code) it wasn't clear to the user what they could/ should do. I felt like this would make supporting contributors difficult. 
- The codeless run time abstracted away a lot of the synthetic data generation process and decision making.  I feel like this is a missed opportunity to share knowledge on synthetic datasets. 

So I have refactored the flow by doing this:
- The app adds the pipeline configuration to the dataset repo which the pipeline uses.
- I removed the codeless runtime and replaced it with a pipeline definition `domain-specific-datasets/distilabel_pipelines/domain_expert_pipeline.py` . 
- The app page supplies a rendered code snippet for how to use this pipeline locally. 
- I exposed  more pipeline configuration for intermediate steps, temperature, and num generations.

burtenshaw commented:
Here's the video guide: https://www.loom.com/share/99f32d7882764d9d8f4dc6ce3d824319?sid=722ead2d-85be-4067-9e0c-b9ebd9864da7
